packages/contracts/contracts/SmartInvoice.sol

```diff
@@ -130,6 +136,35 @@ contract SmartInvoice is
         emit Register(_client, _provider, amounts);
     }
 
+    function addMilestones(uint256[] calldata _milestones) external {
+        require(!locked, "locked");
+        require(block.timestamp < terminationTime, "terminated");
+        require(_msgSender() == client || _msgSender() == provider, "!party");
+        require(_milestones.length > 0, "no milestones are being added");
+        require(_milestones.length <= 10, "only 10 new milestones at a time");
+
+        uint256 newLength = amounts.length + _milestones.length;
+        uint256[] memory baseArray = new uint256[](newLength);
+        uint256 newTotal = total;
+
+        for (uint256 i = 0; i < amounts.length; i++) {
+            baseArray[i] = amounts[i];
+        }
+        for (uint256 i = amounts.length; i < newLength; i++) {
+            baseArray[i] = _milestones[i - amounts.length];
+            newTotal += _milestones[i - amounts.length];
+        }
+
+        total = newTotal;
```

Comment:
I didn't check this with tests yet, but on first glance I believe this is only replacing the total with the total of *only* the new Milestones and doesn't take into account the existing pre-addMilestones() total.  so on 148 initialize to 0 and on 158 total += newTotal.  Can double check later on.

Reply:
newTotal is initialized with the original total value and then adding the new milestone amounts... so either way should result in the same number. It's just a matter of which one is the most gas efficient. It passed testing but I can check to see if your suggestion is even more efficient.

Reply:
Looks like the current method of initializing newTotal = total saves an average of ~100 wei in comparison

Reply:
We can do another pass to optimize if we need to, but if tests are passing I would call it good at this point.

Reply:
Meaning, go ahead and merge and mark complete the related tasks.

Reply:
Ah right; I lapsed that it's initialized with the existing array.